android/engine/src/main/java/org/smartregister/fhircore/engine/cql/LibraryEvaluator.kt

```diff
@@ -0,0 +1,175 @@
+import org.opencds.cqf.cql.evaluator.cql2elm.content.LibraryContentProvider
+import org.opencds.cqf.cql.evaluator.cql2elm.content.fhir.BundleFhirLibraryContentProvider
+import org.opencds.cqf.cql.evaluator.cql2elm.util.LibraryVersionSelector
+import org.opencds.cqf.cql.evaluator.engine.execution.TranslatingLibraryLoader
+import org.opencds.cqf.cql.evaluator.engine.retrieve.BundleRetrieveProvider
+import org.opencds.cqf.cql.evaluator.engine.terminology.BundleTerminologyProvider
+import org.opencds.cqf.cql.evaluator.fhir.adapter.r4.AdapterFactory
+import org.opencds.cqf.cql.evaluator.library.CqlFhirParametersConverter
+import org.opencds.cqf.cql.evaluator.library.LibraryEvaluator
+
+/**
+ * This class contains methods to run CQL evaluators given Fhir expressions It borrows code from
+ * https://github.com/DBCG/CqlEvaluatorSampleApp See also https://www.hl7.org/fhir/
+ */
+class LibraryEvaluator {
+  var fhirContext = FhirContext.forCached(FhirVersionEnum.R4)
+  var adapterFactory = AdapterFactory()
+  var fhirTypeConverter: FhirTypeConverter? = null
+  var cqlFhirParametersConverter: CqlFhirParametersConverter? = null
+  var libraryVersionSelector: LibraryVersionSelector? = null
+  var contentProvider: LibraryContentProvider? = null
+  var terminologyProvider: BundleTerminologyProvider? = null
+  var bundleRetrieveProvider: BundleRetrieveProvider? = null
+  var cqlEvaluator: CqlEvaluator? = null
+  var libraryEvaluator: LibraryEvaluator? = null
+
+  /**
+   * This method loads configurations for CQL evaluation
+   * @param libraryData Fhir resource type Library
+   * @param helperData Fhir resource type LibraryHelper
+   * @param valueSetData Fhir resource type ValueSet
+   * @param testData Fhir resource to evaluate e.g Patient
+   */
+  private fun loadConfigs(
+    libraryData: String,
+    helperData: String,
+    valueSetData: String,
+    testData: String
+  ) {
+    var parser: IParser? = null
+    parser = fhirContext.newJsonParser()
+    parser.setPrettyPrint(true)
+    libraryVersionSelector = LibraryVersionSelector(adapterFactory)
+    fhirTypeConverter = FhirTypeConverterFactory().create(fhirContext.version.version)
+    cqlFhirParametersConverter =
+      CqlFhirParametersConverter(fhirContext, adapterFactory, fhirTypeConverter)
+    // Load Library Content and create a LibraryContentProvider, which is the interface used by the
+    // LibraryLoader for getting library CQL/ELM/etc.
+    val libraryStream: InputStream = ByteArrayInputStream(libraryData.toByteArray())
+    val fhirHelpersStream: InputStream = ByteArrayInputStream(helperData.toByteArray())
+    val library = parser.parseResource(libraryStream)
+    val fhirHelpersLibrary = parser.parseResource(fhirHelpersStream)
+    val resources: List<IBaseResource> = Lists.newArrayList(library, fhirHelpersLibrary)
+    val bundleFactory = fhirContext.newBundleFactory()
+    val bundleLinks = BundleLinks("", null, true, BundleTypeEnum.COLLECTION)
```

Comment:
Is there a way to refactor this into objects that can be created only once when the class is created?

Also do we have an opportunity to maybe refactor and move some of the helper methods to some type of Helper or Util class?

Reply:
@ndegwamartin I can move the load configs method to the class initializer. I also think Helper or Util will be an overkill as they are only used once in this class.

Reply:
It is also for maintainability i.e. better legibility/for team collaboration.